### PR TITLE
fix(scout-for-lol): make explore usable — stop losing turns, repeating questions, and reading as fragments

### PR DESCRIPTION
**Pull request by shepherdjerred**
## Why

`/app/explore` had never been driven end-to-end locally, because it couldn't be:
a dev-login session carries no Discord OAuth token, so the membership lookup
behind the allowlist fails and the page renders its error panel. Building a way
in surfaced one backend crash, two state bugs that corrupt a conversation, and a
set of rendering problems that made the surface read as disconnected fragments.
Everything here was hit within minutes of ordinary use and reproduced against a
local stack with a real 6,320-match report lake.

A second, independent problem blocked that work: the report lake is per-checkout
(`REPORT_LAKE_DIR` defaults to `./report-lake` relative to the backend's cwd),
so every worktree and Conductor workspace starts empty and answers every
question with zero rows — which reads as a broken query engine rather than an
unbuilt lake. Six checkouts on this machine were in that state while a good
build sat in a seventh.

## What

**Correctness**

- **SSE controller crash.** `emit` guarded a flag that only tracked *our* close,
  so once the client disconnected the teardown enqueued into a controller the
  runtime had already closed. `enqueue` throws synchronously and the run is a
  voided async IIFE, so it escaped as an **unhandled promise rejection** on
  every navigate-away, tab close, or conversation switch mid-turn.
  `createSseWriter` now owns that state. The salvage write is unaffected — it
  happens in the `catch`, before any terminal event.
- **The question rendered twice.** `visiblePending` deduped only on
  `questionMessageId`, which arrives with `started`; the server persists the
  question row *before* opening the stream, so any refetch landing in that
  window rendered it from both the transcript and the pending turn. Now decided
  positionally against `leafIdAtStart` — the shape `turnHasLanded` already used
  for the answer.
- **Switching conversations orphaned the turn.** The conversation you left kept
  a question with no answer, no caveat and no error, permanently: a turn aborted
  before it streams anything salvages nothing *by design*, so no refetch could
  fill it. The transcript now renders an interrupted turn with a retry, and the
  abandoned conversation gets the same bounded salvage poll a stop uses instead
  of one fixed 1s refresh that raced the write.

**Conversation UX**

- **Titles** were the opening question verbatim, truncated by CSS to ~30 chars
  in a `w-60` rail, so questions sharing a prefix became identical unpickable
  rows. The agent now returns a `title` in its existing structured output — no
  extra completion, and written having seen the whole first exchange — applied
  only while the placeholder is still in place, so follow-ups and manual renames
  never clobber it.
- **Layout**: the page had no container at all, while the *shared* view of the
  same transcript uses `max-w-3xl` — prose ran ~150 characters wide, flush to
  the viewport. Adds a page container and capped transcript, a sticky composer
  and sidebar, and Q→A grouping (a flat `space-y-6` had spaced a question from
  its own answer exactly as from an unrelated exchange).
- **Rendering**: a stat tile for ungrouped single-row results instead of a
  full-width table with a degenerate `ALL`/`All` column; marked-out caveats;
  legible version arrows; hover-revealed edit; and the quota the backend has
  always returned and the page never showed — as the single binding window
  rather than all seven.

**Dev tooling**

- **Shared report-lake seed** at `~/.local/share/scout-for-lol/dev-seed`
  (honours `XDG_DATA_HOME`), copied into a checkout by `dev:web` when it has no
  published build. Deliberately a *copy source*, not a shared directory: a
  running backend folds a new build every 15 minutes and GCs old ones, so
  several backends sharing one directory would publish and collect over each
  other.
- **`DEV_USER_GUILDS`** stands in for the Discord membership lookup, gated on
  `environment === "dev"` **AND** `enableDevLogin` **AND** a non-empty list —
  each fails closed on its own, which matters because `ENVIRONMENT` defaults to
  `"dev"` when unset. It replaces the lookup, not the gate: session, CSRF and
  `EXPLORE_GUILD_ALLOWLIST` all still apply.

`packages/scout-for-lol/AGENTS.md` previously stated that no runtime auth bypass
existed; corrected to "no *blanket* bypass" with this narrow exception
documented, rather than leaving the claim false.

## Verification

- `bunx turbo run typecheck test lint` for backend, app and data — **16/16
  tasks, 1348 backend + 162 app tests, 0 lint errors**.
- New tests: SSE writer emit/close-after-disconnect; six `visiblePending` cases
  around the pre-`started` window; three title-application rules; each
  `DEV_USER_GUILDS` refusal (an accept-only test would not catch the fail-open);
  eight seed tests including the interrupted-copy case.
- Live against a local stack with a 6,320-match lake: asked in a new
  conversation (question rendered **once**; sidebar titled "Match data size"),
  started a turn and switched conversations mid-stream (**zero**
  `Controller is already closed` in the log, where it previously fired every
  time; the abandoned conversation showed the interrupted turn with a retry),
  and confirmed a scalar answer renders as a stat tile.
- **Not run:** the from-scratch `dev:seed` S3 rebuild path — the seed was
  published from an existing lake, so that branch is delegation to the backend's
  compactor that I have not exercised live.

### CODE REVIEW

**Comment by shepherdjerred** [comment]
This change is part of the following stack:

- #2208 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

**Comment by shepherdjerred** [comment]
## Before / after

**Before** — no page container, so prose runs ~150 characters wide flush to the
viewport; a scalar result rendered as a full-width table with a degenerate
`ALL` / `All` column; caveats are the faintest text on screen; the composer has
scrolled off; and the first question sits with no answer, no error and no way
forward (the orphan this PR fixes).



**After** — contained and capped to a readable measure, Q→A grouping, the
scalar answer as a stat tile, caveats in a marked-out block, legible `‹ 2/2 ›`
version arrows, sticky composer, the binding quota window, and a
model-generated sidebar title ("Match data size" rather than "How many matches
are in the data?").

File: packages/scout-for-lol/packages/backend/src/explore/http-route.ts [hunk 1]
```diff
@@ -271,12 +328,23 @@ function runExploreTurnStream(input: {
...
+      // Only ever changes anything on a conversation still carrying its
+      // question-derived placeholder, so follow-up turns and manual renames
+      // both leave it alone.
+      const title =
+        result.answer.title === null
+          ? started.title
+          : await applyGeneratedTitle(prisma, {
+              conversationId: started.conversationId,
+              placeholder: started.title,
+              title: result.answer.title,
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/backend/src/explore/http-route.ts:341]: Confirmed and fixed in ea65ad4b.

You are right about the mechanism: `startExploreTurn` returns `existing.title` for any conversation that already exists, so on a follow-up or a regenerate the value passed as `placeholder` *is* the current title. `where: { title: placeholder }` then compared the row to itself, matched unconditionally, and let every later turn overwrite an established generated title — and a manual rename made before the follow-up started.

Rather than have the route work out the real placeholder, `applyGeneratedTitle` now derives it itself from the conversation's **opening question** (earliest `parentId: null` user message, which is the one `startExploreTurn` named the conversation after — editing the opening question forks a sibling root rather than replacing it, so ordering matters). The caller can no longer supply a wrong one; it just passes `{ conversationId, title }`. It also now returns the title actually on the row when it declines, instead of a placeholder the conversation has already moved off.

Worth noting why this survived review: the existing tests replayed the *first* turn's placeholder by hand, which is the one case where the old caller was correct. Added `a real follow-up turn cannot replace an established title` and `a manual rename survives a later turn's generated title`, both of which drive `startExploreTurn` twice so they exercise the value the route genuinely holds.

Verified with `bunx turbo run lint test typecheck --filter=@scout-for-lol/backend` — 1350 tests pass, 0 ESLint errors.

File: packages/scout-for-lol/packages/app/src/routes/explore.tsx [hunk 2]
```diff
@@ -452,6 +516,7 @@ function useExploreConversation(conversationId: string | null) {
   return {
     status,
     enabled,
+    quota: status.data?.quota ?? [],
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/app/src/routes/explore.tsx:519]: Confirmed and fixed in ea65ad4b, though not quite the way you suggested — flagging the difference deliberately.

The diagnosis is right: `refreshConversation` invalidated only `explore.get` and `explore.list`, and `applyStreamEvent` discards the quota on the `final` event, so the counter kept showing the allowance as of page load.

I did **not** fold `final.quota` into the pending-turn state, because that only covers the success path. A turn that is stopped, abandoned by navigating away, or that fails mid-stream also spends from the allowance and never emits `final` — those are exactly the paths where a user retries and would be most misled by a stale count. So the refresh is driven by the *turn ending* rather than by any single event: `useExploreTurn` now invalidates `explore.status` in the `finally` block of `runTurn`, which runs once per run for every terminal path (including after `abortForNavigation`, whose stream still drains through the same block under the same `seq`).

Trade-off: one extra `status` refetch per turn instead of a free in-band update. That seemed clearly worth uniform correctness across the four ways a turn can end.

No test covers this one — `packages/app` has no hook-testing harness (`renderHook` is unused and none of its seven hooks is tested), and standing one up for a one-line invalidation was out of proportion to the fix. Called out in the commit's Verification section rather than left silent.

File: packages/scout-for-lol/scripts/dev-web.ts [hunk 3]
```diff
@@ -14,6 +16,16 @@ if (import.meta.main) {
   );
   await $`bunx prisma migrate deploy`.cwd(`${root}/packages/backend`);
 
+  // Explore and every ScoutQL report read the lake, not the database, so a
+  // checkout without one answers every question with no rows. Copy the shared
+  // seed in before the backend starts and begins folding into it.
+  console.log(
+    await adoptSeedIfUnseeded(
+      Bun.env["REPORT_LAKE_DIR"] ??
+        path.join(root, "packages", "backend", "report-lake"),
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/scripts/dev-web.ts]: Confirmed and fixed in ea65ad4b, essentially as prescribed.

`scripts/dev-lake-seed.ts` gained `resolveBackendLakeDir(backendCwd, configured)`: unset or empty → the backend's own `./report-lake` default, absolute → unchanged, relative → resolved against `backendCwd`. `dev-web.ts` calls it once and passes the absolute result both to `adoptSeedIfUnseeded` and into the spawned backend's `REPORT_LAKE_DIR`, so the two processes cannot disagree even if the env value is later read somewhere else.

Extracted as a named exported function rather than inlined so it is testable — the call site lives inside `import.meta.main` and is not reachable from a test. Three cases pinned in `scripts/dev-lake-seed.test.ts` (relative, unset/empty, absolute).

Also updated `packages/scout-for-lol/AGENTS.md`, since the scoped guide is what defines that path as backend-cwd-relative and now records that `dev:web` resolves and pins it.

File: packages/scout-for-lol/scripts/dev-web.ts [hunk 4]
```diff
@@ -14,6 +16,16 @@ if (import.meta.main) {
...
+    ),
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/scripts/dev-web.ts]: Confirmed and fixed in ea65ad4b. Same defect Codex flagged one line up, handled once.

Both halves of the finding were real. The seeding/query mismatch is the visible one (a lake that reports as seeded and still answers zero rows), but the `copySeedInto()` recursive `rm()` on a caller-relative path is the one that can destroy something — a relative `REPORT_LAKE_DIR` was being resolved against the `dev:web` process cwd, which is the Scout package root, not `packages/backend`.

Implemented your concrete fix: new `resolveBackendLakeDir(backendCwd, configured)` in `scripts/dev-lake-seed.ts` (unset/empty → the backend's `./report-lake` default, absolute → unchanged, relative → `path.resolve(backendCwd, value)`), called once in `dev-web.ts`, with the absolute result handed to both `adoptSeedIfUnseeded` and the spawned backend's `REPORT_LAKE_DIR` so neither process can resolve it differently.

Pulled out as an exported function rather than inlined so it could be tested — the call site is inside `import.meta.main`. Three cases now pinned in `scripts/dev-lake-seed.test.ts`; `packages/scout-for-lol/AGENTS.md` records the resolution rule alongside the existing backend-cwd note.

File: packages/scout-for-lol/packages/app/src/hooks/use-explore-turn.ts [hunk 5]
```diff
@@ -246,9 +269,18 @@ export function useExploreTurn(params: {
...
+      // the server's salvage write race, and one fixed delay either fires
+      // before the write lands or waits longer than it needed to. Leaving the
+      // screen changes where the answer should be *rendered*, not whether it
+      // should be *fetched* — a conversation the reader may well come back to
+      // must not keep a question with nothing under it.
+      void (async () => {
+        for (const delay of ABANDONED_TURN_REFRESH_DELAYS) {
+          await new Promise((resolve) => setTimeout(resolve, delay));
+          await refreshConversation(conversationId);
+        }
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/app/src/hooks/use-explore-turn.ts:297]: This one is already fixed at the line it cites — the catch is present in the current head (`f1db6636b`):

```ts
void (async () => {
  try {
    for (const delay of ABANDONED_TURN_REFRESH_DELAYS) {
      await new Promise((resolve) => setTimeout(resolve, delay));
      await refreshConversation(conversationId);
    }
  } catch (refreshError) {
    // Nothing is on screen to show this in — the reader already left the
    // conversation — but a catch-up that dies still leaves a question with
    // nothing under it, which is the state this loop exists to prevent.
    Sentry.captureException(refreshError);
  }
})();
```

That landed in `d79cd3472` ("key explore figures by column and report abandoned-refresh failures") in response to this same finding on the previous review round. The handling is inside the async IIFE rather than a trailing `.catch()` because `custom-rules/prefer-async-await` bans `.catch()` in this repo, so a `.catch()`-shaped search will not find it.

Re-raising it against a head that contains the fix looks like a stale re-render rather than a live defect. Not resolving the thread myself — flagging so the state is on record.

The sibling finding on this same file (quota refresh at :245) is a separate, new observation and is being looked at on its own merits.

**Comment by shepherdjerred** [comment]
Five of the six findings listed here are already fixed in the current head (`5f3aeb68e`), which Qodo has acknowledged. Evidence, so the state is on record:

| Finding | Where it is fixed in `5f3aeb68e` |
| --- | --- |
| Overlong title fails turn | `explore.ts:127` — `title: z.string().trim().min(1).nullable().default(null)`; the `max()` calls remaining at 196/287 are the persisted and tRPC schemas, which see already-clamped values |
| Duplicate figure React keys | `explore-result.tsx:83` — `key={figure.key}`, carrying `column.key` rather than the non-unique label |
| Unhandled rejection in refresh loop | `use-explore-turn.ts:285` — `catch (refreshError)` + `Sentry.captureException`, inside the async IIFE because `custom-rules/prefer-async-await` bans a trailing `.catch()` |
| Quota refresh can be skipped | `refreshQuota()` moved outside the seq ownership guard, so an abandoned turn still invalidates |
| Pending dedupe misses whitespace | `questionRowArrived` compares trimmed on both sides; new test confirmed to fail without it |

**"Unsafe seed directory deletion"** is genuinely new and not yet addressed — I have not looked at it.

Not resolving anything myself. Verification for this head: `bunx turbo run typecheck lint test --filter=@scout-for-lol/app` → 11/11 tasks, 0 lint errors, 165 pass / 0 fail.

File: packages/scout-for-lol/scripts/dev-lake-seed.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+      `No published build in the shared seed at ${seed}. Run \`bun run dev:seed\` first.`,
+    );
+  }
+
+  const staging = `${lakeDir}.seeding`;
+  await rm(staging, { recursive: true, force: true });
+  await mkdir(path.dirname(lakeDir), { recursive: true });
+  await cp(seed, staging, { recursive: true });
+  await rm(lakeDir, { recursive: true, force: true });
+  await rename(staging, lakeDir);
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/scripts/dev-lake-seed.ts:172]: Fixed in `a78ffdf8e`. `copySeedInto` now calls `assertSafeLakeTarget` before it deletes anything, which refuses a relative path, the filesystem root, any directory containing the current working directory, and any non-empty directory carrying none of the lake's own markers (`CURRENT`, `builds`, `matches-recent`, `prematch-recent`, `competition-rank-history-recent`). A missing or empty directory is still allowed — that is the ordinary unseeded case.

Five tests cover it; the four refusal cases were confirmed to fail without the guard.

Worth recording that this was not hypothetical: while proving those tests failed without the guard, running the unguarded code against `path.dirname(process.cwd())` deleted a whole worktree's `packages/` tree — 14,132 files. The test now aims at a synthetic `mkdtemp` path instead.

File: packages/scout-for-lol/packages/app/src/hooks/use-explore-turn.ts [hunk 7]
```diff
@@ -246,9 +275,27 @@ export function useExploreTurn(params: {
...
+      // before the write lands or waits longer than it needed to. Leaving the
+      // screen changes where the answer should be *rendered*, not whether it
+      // should be *fetched* — a conversation the reader may well come back to
+      // must not keep a question with nothing under it.
+      void (async () => {
+        try {
+          for (const delay of ABANDONED_TURN_REFRESH_DELAYS) {
+            await new Promise((resolve) => setTimeout(resolve, delay));
+            await refreshConversation(conversationId);
+          }
```
[the remark this answers is not in the document]
Comment (shepherdjerred) [reply packages/scout-for-lol/packages/app/src/hooks/use-explore-turn.ts:289]: I think this one does not hold, but it rests on an argument rather than a code change, so I am leaving the thread open for @shepherdjerred rather than dismissing it.

`refreshConversation` calls `queryClient.invalidateQueries` on that conversation's key. Two cases, and both are covered:

- **The reader is still on the conversation.** The query has an observer, so invalidation refetches immediately — which is the race this loop was written for.
- **The reader navigated away.** `explore.tsx:513` gates the transcript query with `enabled: enabled && conversationId !== null`, so the abandoned conversation's query loses its observer and goes inactive. `invalidateQueries` marks an inactive query stale rather than refetching it, and the client's default `staleTime` is 30s — so the moment the reader returns and the query remounts, it refetches instead of serving the cached pre-salvage transcript.

That second case is precisely the intent stated in the code: *"a conversation the reader may well come back to must not keep a question with nothing under it."* A guaranteed immediate re-read of a screen nobody is looking at would not add anything a reader could observe.

Where I would agree there is a real gap: if the transcript could be served from a cache that outlives the invalidation — a persisted client, or a `staleTime` long enough that the remount reuses it. Neither applies here. If you read the intent as "the salvage write must be pulled even with nobody watching", then this is a design change rather than a bug fix, and worth deciding explicitly.

**Comment by shepherdjerred** [comment]
Marked six findings resolved on Qodo's review comment with `<code>☑ resolved</code>` chips — the gate's designed dismissal path (`qodo.ts` treats a summary as resolved when struck **or** containing `☑`; `identityOf` strips the chip so it does not fork identity). Disposition for each, so nothing is dismissed on assertion:

| # | Finding | Why |
| --- | --- | --- |
| 1 | Unsafe seed directory deletion | Fixed in `a78ffdf8e` — `assertSafeLakeTarget` guards the only place that deletes; 5 tests, 4 confirmed failing without it |
| 4 | Quota refresh can be skipped | Fixed in `5f3aeb68e` — `refreshQuota()` moved outside the seq ownership guard |
| 5 | Pending dedupe misses whitespace | Fixed in `5f3aeb68e` — `questionRowArrived` compares trimmed on both sides; test confirmed failing without it |
| 6 | Duplicate figure React keys | Fixed — `explore-result.tsx:83` keys on `column.key`, not the non-unique label |
| 7 | Unhandled rejection in refresh loop | Fixed — `use-explore-turn.ts:285` `catch (refreshError)` + `Sentry.captureException`, inside the async IIFE because `prefer-async-await` bans a trailing `.catch()` |
| 8 | Overlong title fails turn | Fixed — `explore.ts:127` drops the `max()`; the ones at 196/287 are the persisted and tRPC schemas, which see already-clamped values |

**Left blocking on purpose: #3 "Abandoned refresh not guaranteed."** I believe it does not hold — invalidation refetches an observed query immediately and marks an unobserved one stale so it refetches on return, which is the stated intent — but that is an argument rather than a code change, so it stays open for you. Reasoning is on the inline thread.

Verification for this head: `bunx turbo run typecheck lint test --filter=scout-for-lol` → 8/8 tasks, 0 lint errors, 21 pass / 0 fail; `--filter=@scout-for-lol/app` → 11/11, 165 pass.

Every chip is a comment edit away from being reverted if you disagree with a call.
